lib/lambda_ethereum_consensus/validator/validator_set.ex

```diff
@@ -0,0 +1,284 @@
+defmodule LambdaEthereumConsensus.ValidatorSet do
+  @moduledoc """
+  Module that holds the set of validators and their states,
+  it also manages the validator's duties as bitmaps to
+  simplify the delegation of work.
+  """
+
+  defstruct head_root: nil, duties: %{}, validators: []
+
+  require Logger
+
+  alias LambdaEthereumConsensus.StateTransition.Misc
+  alias LambdaEthereumConsensus.Validator
+  alias LambdaEthereumConsensus.Validator.Duties
+
+  @type validators :: %{Validator.index() => Validator.t()}
+
+  @type t :: %__MODULE__{
+          head_root: Types.root() | nil,
+          duties: %{
+            Types.epoch() => %{
+              proposers: Duties.proposer_duties(),
+              attesters: Duties.attester_duties()
+            }
+          },
+          validators: validators()
+        }
+
+  @doc "Check if the duties for the given epoch are already computed."
+  defguard is_duties_computed(set, epoch)
+           when is_map(set.duties) and not is_nil(:erlang.map_get(epoch, set.duties))
+
+  @doc """
+  Initiate the set of validators, given the slot and head root.
+  """
+  @spec init(Types.slot(), Types.root()) :: t()
+  def init(slot, head_root) do
+    config = Application.get_env(:lambda_ethereum_consensus, __MODULE__, [])
+    keystore_dir = Keyword.get(config, :keystore_dir)
+    keystore_pass_dir = Keyword.get(config, :keystore_pass_dir)
+
+    setup_validators(slot, head_root, keystore_dir, keystore_pass_dir)
+  end
+
+  defp setup_validators(_s, _r, keystore_dir, keystore_pass_dir)
+       when is_nil(keystore_dir) or is_nil(keystore_pass_dir) do
+    Logger.warning(
+      "[Validator] No keystore_dir or keystore_pass_dir provided. Validators won't start."
+    )
+
+    %__MODULE__{}
+  end
+
+  defp setup_validators(slot, head_root, keystore_dir, keystore_pass_dir) do
+    validator_keystores = decode_validator_keystores(keystore_dir, keystore_pass_dir)
+    epoch = Misc.compute_epoch_at_slot(slot)
+
+    validators =
+      Map.new(validator_keystores, fn keystore ->
+        validator = Validator.new(keystore, slot, head_root)
```

Comment:
Instead of fetching the state internally here, can we fetch the state once in line 57?

Reply:
I'm sure I had that in place at some point in time and then moved back to do it in the Validator, but yeah, I cant think of a good reason now to not move it out, I'll do it!

Reply:
It was my fault, I even had the other new signature in validation without using it, great catch, already solved!